src/auth/baseexternalclient.ts

```diff
@@ -0,0 +1,487 @@
+  /**
+   * @return A promise that resolves with the project ID corresponding to the
+   *   current workload identity pool. When not determinable, this resolves with
+   *   null.
+   *   This is introduced to match the current pattern of using the Auth
+   *   library:
+   *   const projectId = await auth.getProjectId();
+   *   const url = `https://dns.googleapis.com/dns/v1/projects/${projectId}`;
+   *   const res = await client.request({ url });
+   *   The resource may not have permission
+   *   (resourcemanager.projects.get) to call this API or the required
+   *   scopes may not be selected:
+   *   https://cloud.google.com/resource-manager/reference/rest/v1/projects/get#authorization-scopes
+   */
+  async getProjectId(): Promise<string | null> {
+    if (this.projectId) {
+      // Return previously determined project ID.
+      return this.projectId;
+    } else if (this.projectNumber) {
+      // Preferable not to use request() to avoid retrial policies.
+      const headers = await this.getRequestHeaders();
+      const response = await this.transporter.request<ProjectInfo>({
+        headers,
+        url: `${CLOUD_RESOURCE_MANAGER}${this.projectNumber}`,
+        responseType: 'json',
+      });
+      this.projectId = response.data.projectId;
+      return this.projectId;
+    }
+    return null;
+  }
+
+  /**
+   * Authenticates the provided HTTP request, processes it and resolves with the
+   * returned response.
+   * @param opts The HTTP request options.
+   * @param retry Whether the current attempt is a retry after a failed attempt.
+   * @return A promise that resolves with the successful response.
+   */
+  protected async requestAsync<T>(
+    opts: GaxiosOptions,
+    retry = false
+  ): Promise<GaxiosResponse<T>> {
+    let r2: GaxiosResponse;
+    try {
+      const r = await this.getRequestHeaders();
+      opts.headers = opts.headers || {};
+      if (r && r['x-goog-user-project']) {
+        opts.headers['x-goog-user-project'] = r['x-goog-user-project'];
+      }
+      if (r && r.Authorization) {
+        opts.headers.Authorization = r.Authorization;
+      }
+      r2 = await this.transporter.request<T>(opts);
+    } catch (e) {
+      const res = (e as GaxiosError).response;
+      if (res) {
+        const statusCode = res.status;
+        // Retry the request for metadata if the following criteria are true:
+        // - We haven't already retried.  It only makes sense to retry once.
+        // - The response was a 401 or a 403
+        // - The request didn't send a readableStream
+        // - forceRefreshOnFailure is true
+        const isReadableStream = res.config.data instanceof stream.Readable;
+        const isAuthErr = statusCode === 401 || statusCode === 403;
+        if (
+          !retry &&
+          isAuthErr &&
+          !isReadableStream &&
+          this.forceRefreshOnFailure
+        ) {
+          await this.refreshAccessTokenAsync();
+          return this.requestAsync<T>(opts, true);
```

Comment:
Can we do `return await` here? I understand it will have the same behavior in most cases, but I'm concerned about some special case where the exception being thrown will be be caught properly.

Reply:
Done